### DISCUSSION TITLE
Suggestion for default property assignment

### DISCUSSION THREAD

**Discussion by elmar-de-groot** [dotnet/csharplang#2001] [Language Ideas] [answered]
I'd like to make a C# language suggestion. I'm really not sure if this is the right forum to voice this, if not pls point me in the right direction.

What I'd like to see is an enhancement to the property initialization that allows you to return an instance of the requested type when the underlying value is null.

`public IList<int> Numbers {get; set;} ?= new List<int>()`

Note the use of the `?=` assignment. The above would be the equivalent of the following code:

`public IList<int> Numbers {`
   ` get => return _numbers ?? (_numbers = new List<int>();`
   ` set => _numbers = value;`
`)`
`private IList<int> _numbers`

The current auto-property functionality uses the assignment as part of the initialization of the class as if you created the assignment inside the constructor of the class. With the improvements to reduce 'null-ref-exceptions', I feel this could be a great addition and reduce some of the repetitive code.

Regards,

**Comment by YairHalberstadt** [comment]
This is indeed the correct place to make c# language requests.

**Comment by ufcpp** [comment]
similar to https://github.com/dotnet/csharplang/issues/1045, https://github.com/dotnet/csharplang/issues/312

**Comment by quinmars** [comment]
With my favorite proposal #140 this could be written that way:

```csharp
public IList<int> Numbers
{
    set;
    get => field ?? (field = new List<int>());
}
```

**Comment by alrz** [comment, marked as answer]
null-coalescing assignment is a candidate for C# 8.0, so
```cs
public IList<int> Numbers 
{
    set; get => field ??= new List<int>();
}
```
and if the property type is the same as the instance `get => field ??= new();` should work too.

**Comment by quinmars** [comment]
Target-typed new-expressions will come with C# 8.0, so

```csharp
public List<int> Numbers 
{
    set; get => field ??= new();
}
```
(I know, it will not work for the exact OP example)

**Comment by elmar-de-groot** [comment]
Thank you all for your feedback. Looking forward to C# 8.0 :)

**Comment by jnm2** [comment]
Yes, the `field` keyword is the ticket. I hope it's considered for inclusion in some version of C#.
